doc/developer/design/20240609_error_handling.md

```diff
@@ -0,0 +1,302 @@
+stream.
+
+So, to start, we will permit applying the `IGNORE ERRORS` modifier only to
+sources and subsources.
+
+In the future, we can consider expanding the set of relations to which `IGNORE
+ERRORS` can be applied. For example, applying `IGNORE ERRORS` to a SQL query
+which contains only map, filter, and project nodes in its `HirScalarExpr` likely
+has sensible semantics under `IGNORE ERRORS`: the computation proceeds row by
+row, and any errors while applying the computation to a row omits only that one
+row from the output.
+
+#### `ONLY ERRORS`
+
+There are two important guardrails to apply to `ONLY ERRORS`:
+
+1. We will permit applying the `ONLY ERRORS` modifier only to persist-backed
+   objects: tables, sources, and materialized views. This relieves some pressure
+   on the compute layer, which already does not assume that the errors in a
+   persist shard have any particular relationship to their source.
+
+2. We will clearly document that error messages and codes are **not stable**. We
+   reserve the right to reword error messages, change error codes, add new
+   errors, or eliminate errors in future releases of Materialize.
+
+   Note also that `ONLY ERRORS` doesn't fundamentally add new backwards
+   compatibility surface area. Error messages, descriptions, hints, and codes
+   are already returned via the SQL interface, and users may already be
+   erroneously relying on seeing particular error messages and codes from
+   Materialize in particular situations.
+
+### Implementation
+
+#### `IGNORE ERRORS`
+
+[@maddyblue] has a prototype implementation of `IGNORE ERRORS` in [#27137].
+The core of the implementation is the addition of a new `ignore_errors`
+field to `MirRelationExpr::Get`:
+
+```diff
+ MirRelationExpr::Get {
+     /// A global or local identifier naming the collection.
+     id: Id,
+
+     // ...
+
++    /// Whether to ignore errors.
++    ignore_errors: bool,
+ },
+```
+
+When rendered, if `ignore_errors` is true, the referenced collection's errs
+stream is simply not wired up to the downstream error stream.
+
+**Open questions:**
+
+  * Is this implementation acceptable to the cluster team?
+  * Will the existence of the new field cause correctness issues if existing
+    references to `MirRelationExpr` fail to check the field?
+
+#### `ONLY ERRORS`
+
+There is not yet a prototype for `ONLY ERRORS`.
+
+One candidate proposal is to introduce a `FlavoredGid` type, which couples a
+global ID with a specification of whether the oks stream or the errs stream is
+of interest.
+
+```diff
++enum GidFlavor {
++    Oks,
++    Errs,
++}
++
++struct FlavoredGid {
++    id: GlobalId,
++    flavor: GidFlavor,
++}
++
+ pub enum Id {
+     Local(LocalId),
+-    Global(GlobalId),
++    Global(FlavoredGid),
+ }
+```
+
+This would be an unfortunately massive change. We'd need to audit every use of
+`GlobalId` across the codebase to assess whether it should remain a `GlobalId`
+or become a `FlavoredGid`.
```

Comment:
@frankmcsherry is deeply suspicious of that approach, and for good reason. The problem is that a collection is would no longer be identified by just `GlobalId` but `(GlobalId, flavor)`. So anywhere that is e.g. storing a `GlobalId` in a map is potentially wrong because the map key really needs to be `(GlobalId, flavor)`.

Another possibly less stressful way to tackle `ONLY ERRORS` would be to allocate _two_ global IDs for each persist-backed collection, one representing the oks stream and one representing the errs stream. Then the optimizer and compute layers could be essentially oblivious to the existence of this feature. It would introduce stress in other components, though—for example neither the catalog nor the storage controller are prepared to handle multiple global IDs per {table, source, matview}.

Reply:
> The problem is that a collection is would no longer be identified by just `GlobalId` but `(GlobalId, flavor)`. So anywhere that is e.g. storing a `GlobalId` in a map is potentially wrong because the map key really needs to be `(GlobalId, flavor)`.

Hmm, looking through the usage of `GlobalId` in Compute, this would be indeed more work than I initially thought. But at least, in every specific instance that I checked, the change seems straightforward. It's just that there are quite a lot of them.

> Another possibly less stressful way to tackle ONLY ERRORS would be to allocate two global IDs for each persist-backed collection, one representing the oks stream and one representing the errs stream. Then the optimizer and compute layers could be essentially oblivious to the existence of this feature. It would introduce stress in other components, though—for example neither the catalog nor the storage controller are prepared to handle multiple global IDs per {table, source, matview}.

Yeah, this would certainly make it much easier for Compute. But I'm not sure which would be the easier approach overall.

Reply:
One more possible approach: How about we change to `FlavoredGid` only in Compute? 
- Adapter could keep using the normal `GlobalId` for its own purposes.
- When Adapter communicates with Compute, it would translate between `GlobalId` and `FlavoredGid`.
- Inside Compute, changing all occurrences of `GlobalId` to `FlavoredGid` might be easier than changing many occurrences of `GlobalId` to `(GlobalId, Flavor)`. This is because in many cases `GlobalId`s are used as a black box, and with a `FlavoredGid` this black box would just change to a different black box, while with `(GlobalId, Flavor)` the flavor wouldn't be hidden.

Reply:
So, a little summary of the possible approaches from above:
1. `(GlobalId, Flavor)` in Compute.
"The problem is that a collection would no longer be identified by just `GlobalId` but `(GlobalId, Flavor)`. So anywhere that is e.g. storing a `GlobalId` in a map is potentially wrong because the map key really needs to be `(GlobalId, flavor)`."
2. `FlavoredGid` only in Compute.
3. Two `GlobalId`s for each collection, at the Adapter level.
4. `FlavoredGid` everywhere. This would be a really massive change. My IDE says there are 1741 occurrences of GlobalId.

Reply:
That seems like a good list!

Reply:
@ggevay thank you for the detailed review and accompanying code analysis!
@ggevay @benesch @frankmcsherry @teskje @antiguru would it work to move forward with option 2 then? It sounds like the sweetspot.

Reply:
@bosconi I think it'd be good to get a live design review meeting on the books with the set of people you tagged to discuss the tradeoffs. We should make sure to give the Kafka source-specific DLQ alternative some airtime too. IMO it trades of tech debt for product debt, which may be the right call here.

@JLDLaughlin may want to attend as well to weigh in on the product considerations.

Reply:
I have in the past been thinking of having the compute layer (and maybe storage too) identify collections by a `CollectionId` that's distinct from `GlobalId`. That would decouple adapter's object identifiers from compute's collection identifiers and would have been helpful in the past in at least two instances:

* `ALTER MATERIALIZED VIEW ... SET CLUSTER ...`, where compute may need to have multiple dataflows installed for the same materialized view at the same time.
* Unified compute introspection, where compute would like to mint its own collection IDs, which is made difficult because `GlobalId` generation is owned by adapter.

It seems like a `CollectionId` would also help in this case. So my suggestion would be to use the `FlavoredGid` approach but use a better name and make sure it remains extensible, so probably an enum rather than a struct.

> 3. Two `GlobalId`s for each collection, at the Adapter level.

Having the option to store multiple `GlobalId`s for a catalog object would be great also because it'd allow us to give different IDs to the two parts comprising a materialized view (storage collection and dataflow). In the past we had sometimes [bugs](https://github.com/MaterializeInc/materialize/issues/27484) caused by them having the same ID, which leads to confusion particularly in frontier management.

Reply:
> I have in the past been thinking of having the compute layer (and maybe storage too) identify collections by a `CollectionId` that's distinct from `GlobalId`. That would decouple adapter's object identifiers from compute's collection identifiers and would have been helpful in the past in at least two instances:

Would `REPLAN MATERIALIZED VIEW` / `REPLAN INDEX` also benefit from this in the same way as `ALTER MATERIALIZED VIEW ... SET CLUSTER ...`?

Sounds interesting, but this seems to be a bigger change than what I was thinking for 2.: I'm hoping that the Controller can keep using the original `GlobalId`s. So, `FlavoredGid`s (or its new name) would be used by plans and `DataflowDescription` and the things that directly manipulate `DataflowDescription`s, but not in the Controller.